runtime/j9vm/jvm.c

```diff
@@ -4234,95 +4579,206 @@ JVM_NativePath(char* path)
+ *
+ *  @param  sigNum  Integer value of the signal to be sent to the
+ *                  calling process or thread
+
+ *  @returns JNI_TRUE if the signal is successfully raised/sent
+ *           JNI_FALSE if the signal is not sent
+ *
+ *	DLL: jvm
  */
 jboolean JNICALL
 JVM_RaiseSignal(jint sigNum)
 {
+	jboolean rc = JNI_FALSE;
+	J9JavaVM *javaVM = (J9JavaVM *)BFUjavaVM;
 
 	Trc_SC_RaiseSignal_Entry(sigNum);
 
-	/**** TBD :: ignore certain signals when -Xrs is on ****/
-	raise(sigNum);
+	if (J9_IS_SIGNAL_SPECIAL(sigNum)
+			&& J9_ARE_ALL_BITS_SET(javaVM->sigFlags, J9_SIG_XRS)) {
+		/* Ignore signal */
+	} else {
+		raise(sigNum);
+		rc = JNI_TRUE;
+	}
 
-	Trc_SC_RaiseSignal_Exit(JNI_TRUE);
+	Trc_SC_RaiseSignal_Exit(rc);
 
-	return JNI_TRUE;
+	return rc;
 }
 
+/**
+ *  This is a stub for the pre-defined handler. The pre-defined
+ *  handler is supposed to be used in JVM_RegisterSignal when
+ *  the special value of J9_PRE_DEFINED_HANDLER_CHECK (2) is specified
+ *  in the handler. It hasn't been implemented since its functionality
+ *  is not known.
+ *
+ *  TODO: Implement the pre-defined handler.
+ *
+ *  @param  sigNum  Integer value of the signal to be sent to the
+ *                  calling process or thread
+ *
+ *  @returns void
+ *
+ *	DLL: jvm
+ */
+static void
+dummySignalHandler(int sigNum) {
+
+}
 
 /**
- * Method stub for method not yet implemented.
+ *  Register a signal handler for a signal. Signals such as SIGFPE,
+ *  SIGILL and SIGSEGV are used by the VM; so, we don't register
+ *  a signal handler for these signals. If "-Xrs" commandline option
+ *  is specified, then signals such as SIGQUIT, SIGHUP, SIGINT
+ *  and SIGTERM are also ignored; thus, we don't register a signal
+ *  handler for these signals. If handler has the special value of
+ *  J9_PRE_DEFINED_HANDLER_CHECK (2), then handler is changed to
+ *  dummySignalHandler before it is registered. If the old handler
+ *  is same as the new handler, then a special value,
+ *  J9_OLDHANDLER_SAME_AS_NEWHANDLER (2) is returned.
+ *
+ *  @param  sigNum  Integer value of the signal to be sent to the
+ *                  calling process or thread
+ *  @param  handler New handler to be associated to the signal
+ *
+ *  @returns address of the old signal handler on success
+ *           J9_SIG_ERR (-1) in case of error
+ *
+ *	DLL: jvm
  */
 void* JNICALL
 JVM_RegisterSignal(jint sigNum, void* handler)
 {
+	J9JavaVM *javaVM = (J9JavaVM *)BFUjavaVM;
+	void *oldHandler = (void *)J9_SIG_ERR;
+
+#if !defined(WIN32)
+	struct sigaction newSignalAction = {{0}};
+	struct sigaction oldSignalAction = {{0}};
+#endif /* !defined(WIN32) */
 
 	Trc_SC_RegisterSignal();
 
-	/**** TBD :: allow signal handler registration (thru portlib?) ****/
-	return (void *) -1;
+	if (J9_IS_SIGNAL_USED_BY_VM(sigNum)) {
+		/* Don't allow user to register a native handler since
+		 * the signal is already used by the VM.
+		 */
+	} else {
+		if (J9_IS_SIGNAL_SPECIAL(sigNum)
```

Comment:
`elseif` can be used here to reduce indentation

Reply:
fixed.